[dnf5] rpm::SolvQuery inherits rpm::PackageSet, improve SolvMap and PackageSet

jrohel opened the pull request:
Improved rpm :: sol :: SolvMap and rpm :: PackageSet:
- added swap() method
- fixed and optimized the copy assignment operator and the move assignment operator

Reorganized `rpm::SolvQuery`. It now inherits `rpm::PackageSet`. It support iteration. Iterators are inherited from `rpm::PackageSet`.

Unit tests and applications use `rpm::SolvQuery` directly instead of `rpm::PackageSet` helping object .

jrohel committed:
rpm::solv::SolvMap: add swap() method, add noexcept to methods

jrohel committed:
rpm::solv::PackageSet: add swap() method, add noexcept to methods

jrohel committed:
rpm::SolvQuery: contains PackageSet instead of SolvSackWeakPtr and SolvMap

jrohel committed:
rpm::SolvQuery inherits rpm::PackageSet

jrohel committed:
rpm::PackageSet: Add copy and move assignment operators, improve move ctor

Adds:
PackageSet & operator=(const PackageSet & src);
PackageSet & operator=(PackageSet && src);

The move constructor preserves the pImpl in the source PackageSet. ->
The methods of source PackageSet object do not refer to nullptr.

jrohel committed:
rpm::solv::SolvMap: Improve copy and move assignment operators

Optimized the copy assignment operator:
- Added check for self-assignment.
- If possible, reuse storage (releases and allocates new ones only if
  operand size differs)

Optimized/fixed the move assignment operator:
- Added check for self-assignment.
- Instead of swap(), copy the size and poiter and reset the sorce.

jrohel committed:
tests: iterate directly through SolvQuery instead of PackageSet

The rpm::SolvQuery inherits rpm::PackageSet. The iterators are inherited
as well. So rpm::SolvQuery can be iterated.

jrohel committed:
Use rpm::SolvQuery directly instead of rpm::PackageSet

Rpm::SolvQuery inherits rpm::PackageSet. So rpm::SolvQuery can be used
as rpm::PackageSet.

j-mracek commented on dnfdaemon-server/services/repo/repo.cpp:
```diff
@@ -49,7 +49,7 @@ uint64_t repo_size(libdnf::rpm::SolvSack & sack, const libdnf::WeakPtr<libdnf::r
     libdnf::rpm::SolvQuery query(&sack);
     std::vector<std::string> reponames = {repo->get_id()};
-    auto pkgset = query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames).get_package_set();
+    auto pkgset = query.ifilter_repoid(libdnf::sack::QueryCmp::EQ, reponames);
```
Why do you want to create an additional reference? What about to use query directly for iteration?
```
for (auto pkg : query) {
```

jrohel replied:
Yes. Good point. I overlooked it.

jrohel replied:
Removed.

j-mracek commented on include/libdnf/rpm/solv_query.hpp:
```diff
@@ -116,7 +115,8 @@ class SolvQuery {
     /// @replaces libdnf/sack/query.hpp:method:Query.run()
-    PackageSet get_package_set() const;
+    PackageSet & get_package_set();
+    const PackageSet & get_package_set() const;
```
Will we need it at all?

jrohel replied:
No. I removed it.

j-mracek commented on include/libdnf/rpm/solv_query.hpp:
```diff
@@ -358,10 +358,12 @@ class SolvQuery {
     /// @replaces libdnf/sack/query.hpp:method:Query.size()
-    std::size_t size() const noexcept;
+    // inherited from PackageSet
+    // std::size_t size() const noexcept;
```
Are you sure that the doc string will work correctly?

jrohel replied:
I will remove the lines.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -36,7 +35,7 @@ void add_obseletes(const libdnf::rpm::SolvQuery & base_query, libdnf::rpm::Packa
     libdnf::rpm::SolvQuery obsoletes_query(base_query);
     obsoletes_query.ifilter_obsoletes(libdnf::sack::QueryCmp::EQ, data);
     // TODO(jmracek) Replace obsoletes_query.get_package_set(); by more effective approach
```
Please could you remove TODO? It is resolved by your change

jrohel replied:
OK. I will remove.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -406,13 +405,13 @@ void Goal::Impl::add_upgrades_to_goal() {
             // TODO(jmracek) Replace obsoletes_query.get_package_set(); by more effective approach
```
Please could you remove `// TODO(jmracek) Replace obsoletes_query.get_package_set(); by more effective approach`. It is already resolved.

jrohel replied:
OK. I will remove.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -406,13 +405,13 @@ void Goal::Impl::add_upgrades_to_goal() {
+            obsoletes_query.ifilter_obsoletes(libdnf::sack::QueryCmp::EQ, query);
             // obsoletes = self.sack.query().available().filterm(obsoletes=installed_query.union(q.upgrades()))
-            query.p_impl->query_result |= obsoletes_query.p_impl->query_result;
+            *query.pImpl |= *obsoletes_query.pImpl;
```
What about to make it even more simple like `query |= obsoletes_query;`?

jrohel replied:
It was an optimization to bypass the sack test. But I'll change it. The code will be more readable.

j-mracek commented on libdnf/rpm/solv/solv_map.hpp:
```diff
@@ -150,24 +152,42 @@ inline SolvMap::SolvMap(int size) {
 
+inline SolvMap & SolvMap::operator=(const SolvMap & other) noexcept {
+    if (this != &other) {
+        if (map.size == other.map.size) {
```
Nice :+1:

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -194,98 +194,66 @@ Id what_downgrades(Pool *pool, Solvable * solvable) {
+            // TODO(jmracek) add exclude application
+        case InitFlags::APPLY_EXCLUDES:
+        case InitFlags::IGNORE_MODULAR_EXCLUDES:
+        case InitFlags::IGNORE_REGULAR_EXCLUDES:
```
In future (hope near) we will need to store InitFlags flags (one if ifilters has a TODO). May I ask you to implement it?

jrohel replied:
I added new commit with implementation.

j-mracek commented on libdnf/rpm/solv/solv_map.hpp:
```diff
@@ -150,24 +152,42 @@ inline SolvMap::SolvMap(int size) {
 inline SolvMap::SolvMap(const SolvMap & other) : SolvMap(other.get_map()) {}
 
 inline SolvMap::SolvMap(SolvMap && other) noexcept {
-    map.size = other.map.size;
```
I suggest that you have to copy a map size `map.size = other.map.size;` for a correct functionality.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -194,98 +194,66 @@ Id what_downgrades(Pool *pool, Solvable * solvable) {
 SolvQuery & SolvQuery::operator|=(const SolvQuery & other) {
-    if (p_impl->sack != other.p_impl->sack) {
-        throw UsedDifferentSack("Cannot perform the action with SolvQuery instances initialized with different SolvSacks");
+    if (pImpl->sack != other.pImpl->sack) {
```
Already discussed - Please could you move the check to PackageSet?

jrohel replied:
I added commit which move the check to PackageSet.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -194,98 +194,66 @@ Id what_downgrades(Pool *pool, Solvable * solvable) {
 SolvQuery & SolvQuery::operator&=(const SolvQuery & other) {
-    if (p_impl->sack != other.p_impl->sack) {
-        throw UsedDifferentSack("Cannot perform the action with SolvQuery instances initialized with different SolvSacks");
+    if (pImpl->sack != other.pImpl->sack) {
```
Already discussed - Please could you move the check to PackageSet?

jrohel replied:
I added commit which move the check to PackageSet.

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -424,22 +395,22 @@ inline static void filter_evr_internal(
 
 SolvQuery & SolvQuery::ifilter_evr(libdnf::sack::QueryCmp cmp_type, const std::vector<std::string> & patterns) {
-    Pool * pool = p_impl->sack->pImpl->get_pool();
+    Pool * pool = pImpl->sack->pImpl->get_pool();
```
May I ask you for rename or PackageSet `pImpl` to `p_impl`? Or please set up a TODO. Just to unify an internal interface.

jrohel replied:
I added commit that unifies it in all classes (PackageSet, PackageSetIterator, ReldepList, ReldepListIterator, SolvSack)

j-mracek commented on libdnf/rpm/solv_query.cpp:
```diff
@@ -1581,6 +1572,8 @@ void SolvQuery::Impl::filter_nevra(
         tmp_cmp_type = (tmp_cmp_type - libdnf::sack::QueryCmp::GLOB) | libdnf::sack::QueryCmp::EQ;
     }
 
+    Pool * pool = pkg_set.get_sack()->pImpl->get_pool();
```
I have to think about (note for reviewer)

jrohel committed:
Move sacks equality tests from SolvQuery to PackageSet

jrohel committed:
rpm::SolvQuery: store init_flags

Init_flags will be needed for some filters.

jrohel committed:
rpm::SolvQuery: remove get_package_set() - SolvQuery can be used directly

Solquery inherits from PackageSet. And it can be used directly instead
of PackageSet.

jrohel committed:
Unification: Rename - "p_impl" in all places

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -132,13 +130,13 @@ void Goal::add_rpm_remove(const libdnf::rpm::Package & rpm_package) {
     }
 }
 
 void Goal::add_rpm_remove(const libdnf::rpm::SolvQuery & query) {
```
Minor - I suggest that the whole function could be removed

jrohel replied:
Thanks. Removed.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -108,15 +106,15 @@ void Goal::add_rpm_install(const libdnf::rpm::Package & rpm_package, bool strict
     p_impl->install_rpm_ids.push_back(std::make_pair(std::move(ids), strict));
 }
 
 void Goal::add_rpm_install(const libdnf::rpm::SolvQuery & query, bool strict) {
```
Minor - I suggest that the whole function could be removed

jrohel replied:
Removed.

j-mracek commented on libdnf/base/goal.cpp:
```diff
@@ -155,23 +153,23 @@ void Goal::add_rpm_upgrade(const libdnf::rpm::Package & rpm_package) {
     p_impl->upgrade_rpm_ids.push_back(std::move(ids));
 }
 
 void Goal::add_rpm_upgrade(const libdnf::rpm::SolvQuery & query) {
```
Minor - I suggest that the whole function could be removed

jrohel replied:
Removed.

j-mracek commented:
LGTM! Only I suggest to delete 3 method because they are not needed any more. But it can be done in another PR.

jrohel committed:
Goal: remove redundant methods that accept rpm::SolvQuery

rpm::SolvQuery inherits rpm::PackageSet. Therefore, methods that accept
rpm :: PackageSet are sufficient.

jrohel commented:
Thanks for review. I removed redundant methods in new commit.
